### PR TITLE
CNTRLPLANE-3999: add declarative TestPlan for composing v2 cluster variants and test matrices

### PR DESCRIPTION
**Pull request by ironcladlou**
Introduce a TestPlan type that decouples variant selection and test matrix arrangement from the platform config Go code. Plans can be defined externally as JSON or YAML and loaded via the TEST_PLAN env var, giving developers and CI jobs a uniform interface for defining custom test workflows without modifying engine code. When unset, engines fall back to the platform's default plan.

ExtraEnv existed only to rename RELEASE_IMAGE_LATEST to E2E_LATEST_RELEASE_IMAGE via the plan; run-tests can do that once at startup like it already does for EVENTUALLY_VERBOSE, keeping env wiring  out of the plan schema.

## Summary by CodeRabbit

- **New Features**
  - Added configurable test plans for selecting specific cluster variants and test suites.
  - Supports test plans defined in JSON or YAML files.
  - Added platform-specific default test plans for AWS and Azure.
  - Validates selected variants, prevents duplicates, and preserves configured order.
  - Generates consistent JUnit report filenames for parallel and sequential tests.

- **Bug Fixes**
  - Improved error reporting for invalid or unresolved test-plan configuration.
  - Added stricter validation for malformed test-plan files and unsupported variants.

### CODE REVIEW

**Comment by ironcladlou** [comment]
cc @joelsmith @mgencur

File: test/e2e/v2/cmd/create-guests/main.go [hunk 1]
```diff
@@ -104,6 +105,7 @@ func loadEnvConfig() envConfig {
 	}
 
 	cfg := envConfig{
+		testPlanPath: os.Getenv("TEST_PLAN"),
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/cmd/create-guests/main.go:108]: the env var catalog is for the test runtime (BeforeSuite context), not the lifecycle CLI tools

File: test/e2e/v2/lifecycle/platform.go [hunk 2]
```diff
@@ -12,35 +12,40 @@ import (
...
+	Variant     string   `json:"variant"`
+	LabelFilter string   `json:"labelFilter"`
+	Skip        string   `json:"skip,omitempty"`
+	ExtraEnv    []string `json:"extraEnv,omitempty"`
+}
+
+// JUnitFile returns the deterministic JUnit XML filename for this
+// test group, derived from the variant and group name.
+func (g TestGroup) JUnitFile() string {
+	return fmt.Sprintf("junit_%s_%s.xml", g.Variant, g.Name)
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/platform.go]: Fixed

File: test/e2e/v2/lifecycle/testplan.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	var errs []error
+
+	// Check that all plan variants exist in the registry.
+	planVariants := make(map[string]bool, len(p.Variants))
+	for _, v := range p.Variants {
+		planVariants[v] = true
+		if !known[v] {
+			errs = append(errs, fmt.Errorf("variant %q not in platform registry", v))
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go]: Fixed

File: test/e2e/v2/lifecycle/testplan.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+func ParseTestPlan(data []byte, yaml bool) (*TestPlan, error) {
+	var plan TestPlan
+	if yaml {
+		if err := sigyaml.UnmarshalStrict(data, &plan); err != nil {
+			return nil, fmt.Errorf("unmarshaling YAML test plan: %w", err)
+		}
+	} else {
+		if err := json.Unmarshal(data, &plan); err != nil {
+			return nil, fmt.Errorf("unmarshaling JSON test plan: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go:95]: Fixed

**Comment by joelsmith** [comment]
The current `TEST_PLAN` env var is a file path, but for CI jobs it would ususally be much easier to pass the plan inline via the job env block with no intermediate step to write a file. We could have `TEST_PLAN` auto-detect inline config and skip the file load, but since this supports both JSON and YAML, I think it might be fragile (since YAML doesn't have a reliable prefix the way JSON's `{` works).

What if you supported `TEST_PLAN` and `TEST_PLAN_FILE` with the former being inline config and the latter being what you currently have.  If both are set, I think it makes sense to have `TEST_PLAN` take precedence.
  
That would let me do something like
```
TEST_PLAN: '{"name":"autonode-standalone-ko","platform":"aws","variants":["karpenter"],"testMatrix":{"parallel":[{"name":"karpenter","variant":"karpenter","labelFilter":"karpenter"}]}}'

```
in the release repo in the env section of my new test. Otherwise, I'd need to figure out some kind of shell wrapper to inject the file into the test container (which isn't super difficult, but is an extra step vs. being able to specify inline config in an env var.

File: test/e2e/v2/lifecycle/testplan.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	if yaml {
+		if err := sigyaml.UnmarshalStrict(data, &plan); err != nil {
+			return nil, fmt.Errorf("unmarshaling YAML test plan: %w", err)
+		}
+	} else {
+		dec := json.NewDecoder(bytes.NewReader(data))
+		dec.DisallowUnknownFields()
+		if err := dec.Decode(&plan); err != nil {
+			return nil, fmt.Errorf("unmarshaling JSON test plan: %w", err)
+		}
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go:95]: I am not going to cover every possible edge case for this parsing, sorry

File: test/e2e/v2/lifecycle/testplan.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+// the platform's default plan if path is empty. Callers are
+// responsible for validation appropriate to their context:
+// create-guests validates against the variant registry, run-tests
+// validates against the cluster manifest.
+func ResolveTestPlan(path string, platform PlatformConfig) (*TestPlan, error) {
+	if path != "" {
+		return ReadTestPlan(path)
+	}
+	plan := platform.DefaultTestPlan()
+	return &plan, nil
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go:79]: Then the user made a mistake and should fix the job definition, I'm not going to cover all these edge cases right now

**Comment by ironcladlou** [comment]
> The current `TEST_PLAN` env var is a file path, but for CI jobs it would ususally be much easier to pass the plan inline via the job env block with no intermediate step to write a file. We could have `TEST_PLAN` auto-detect inline config and skip the file load, but since this supports both JSON and YAML, I think it might be fragile (since YAML doesn't have a reliable prefix the way JSON's `{` works).
> 
> What if you supported `TEST_PLAN` and `TEST_PLAN_FILE` with the former being inline config and the latter being what you currently have. If both are set, I think it makes sense to have `TEST_PLAN` take precedence.
> 
> That would let me do something like
> 
> ```
> TEST_PLAN: '{"name":"autonode-standalone-ko","platform":"aws","variants":["karpenter"],"testMatrix":{"parallel":[{"name":"karpenter","variant":"karpenter","labelFilter":"karpenter"}]}}'
> ```
> 
> in the release repo in the env section of my new test. Otherwise, I'd need to figure out some kind of shell wrapper to inject the file into the test container (which isn't super difficult, but is an extra step vs. being able to specify inline config in an env var.

I understand the need but think this should be handled at a higher later (e.g. Prow) one way or another. A step that lays down the file (as you mentioned) would work but like you said kind of annoying. Another idea would be to bake support into the `hypershift-aws-run-e2e-v2` workflow so you can pass your literal JSON to the workflow in an env var the workflow supports, and the workflow itself could handle writing it out to a file. The latter seems like it would handle the case you want generically, what do you think?

**Comment by joelsmith** [comment]
> I understand the need but think this should be handled at a higher level (e.g. Prow) one way or another. A step that lays down the file (as you mentioned) would work but like you said kind of annoying. Another idea would be to bake support into the `hypershift-aws-run-e2e-v2` workflow so you can pass your literal JSON to the workflow in an env var the workflow supports, and the workflow itself could handle writing it out to a file. The latter seems like it would handle the case you want generically, what do you think?

That seems fine. If the workflow has the feature, that's just as good as the test having the feature, from the PoV of someone creating a new test in the release repo.

File: test/e2e/v2/lifecycle/testplan.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+// TestPlan is a declarative specification of which cluster variants to
+// create and which tests to run against them. It is the external API
+// for composing e2e runs — the variant registry (PlatformConfig) owns
+// the details of how each variant is created, while the TestPlan
+// selects and arranges them.
+type TestPlan struct {
+	Name       string     `json:"name"`
+	Platform   string     `json:"platform"`
+	Variants   []string   `json:"variants"`
```
Comment (csrwng) [inline test/e2e/v2/lifecycle/testplan.go]: This field seems redundant. We can derive it from the TestMatrix.

Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go]: Great idea, even simpler... thank you

Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go]: Implemented, thanks again, good UX improvement and simpler implementation

File: test/e2e/v2/lifecycle/platform.go [hunk 8]
```diff
@@ -6,41 +6,71 @@ import (
...
+// JUnitFile returns the deterministic JUnit XML filename for this
+// test group, derived from the variant and group name. It panics
+// if either component contains path separators or traversal sequences.
+func (g TestGroup) JUnitFile() string {
+	for _, s := range []string{g.Variant, g.Name} {
+		if strings.Contains(s, "/") || strings.Contains(s, "\\") || strings.Contains(s, "..") {
+			panic(fmt.Sprintf("invalid path component in test group: %q", s))
+		}
+	}
+	return fmt.Sprintf("junit_%s_%s.xml", g.Variant, g.Name)
```
[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/platform.go]: We might want to declare and enforce an invariant within TestMatrix that group names should be unique within a TestMatrix (regardless of whether they're parallel or sequential)

[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/platform.go]: If TestGroup.Name is globally unique, and JUnitFile is a function of and always contains TestGroup.Name, how can there ever be a collision of JUnit filenames?

[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/platform.go]: Now I see what you mean and agree it should be addressed

[the remark this answers is not in the document]
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/platform.go]: The latest commit now enforces an invariant that TestGroup names within a TestMatrix must be unique, and Junit files are now based only on the unique TestGroup name (the variant info is irrelevant in this context so nothing important is lost)

File: test/e2e/v2/lifecycle/aws.go [hunk 9]
```diff
@@ -119,20 +119,26 @@ func (a *AWSPlatformConfig) PostVersionRollout(ctx context.Context, cl crclient.
 	return nil
 }
 
-func (a *AWSPlatformConfig) TestMatrix(releaseImage string) TestMatrix {
+func (a *AWSPlatformConfig) DefaultTestPlan() TestPlan {
```
Comment (cblecker) [inline test/e2e/v2/lifecycle/aws.go:122]: The default plan's matrix and `ClusterSpecs()` have to stay in sync by hand — every variant referenced here needs a matching entry in `ClusterSpecs()`, and the `PlatformConfig` docstring says `DefaultTestPlan` selects "all known variants." Nothing enforces that today: the tests validate against a synthetic registry and never construct the real `AWSPlatformConfig`/`AzurePlatformConfig`, so a drift only surfaces once `create-guests` starts provisioning in CI. A small table test ranging over the real platform configs that asserts `plan.Validate(ClusterSpecs(...)) == nil` (and, if the docstring is meant literally, that every spec variant is referenced by the matrix) would catch it in milliseconds. Alternatively, soften the docstring if the default plan is intentionally a curated subset.

File: test/e2e/v2/lifecycle/testplan.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("test plan %q validation failed:\n%w", p.Name, errors.Join(errs...))
+	}
+	return nil
+}
+
+// FilterClusterSpecs returns the subset of allSpecs whose variants
+// are referenced by the TestMatrix.
```
Comment (cblecker) [inline test/e2e/v2/lifecycle/testplan.go]: Nit, non-blocking: `Validate` and `FilterClusterSpecs` both build a variant→registry map and independently report missing variants, with different wording (`variant %q not in platform registry` vs `variants not found in registry: %v`), and `create-guests` calls them back to back. Could have `FilterClusterSpecs` assume a validated plan (or call `Validate` internally) so there's a single source of truth for that check. Fine to leave as-is if you prefer.

File: test/e2e/v2/cmd/run-tests/main.go [hunk 11]
```diff
@@ -63,9 +66,14 @@ func main() {
 	// Let the platform set up any env vars it needs for tests.
 	platform.SetupTestEnv(sharedDir)
 
-	matrix := platform.TestMatrix(releaseImage)
+	testPlanPath := os.Getenv("TEST_PLAN")
+	plan, err := lifecycle.ResolveTestPlan(testPlanPath, platform)
+	if err != nil {
+		log.Fatalf("Failed to resolve test plan: %v", err)
+	}
+	log.Printf("Using test plan %q", plan.Name)
```
Comment (cblecker) [inline test/e2e/v2/cmd/run-tests/main.go:74]: `run-tests` resolves the plan here but never calls `plan.Validate()` — only `ResolveVariants` runs, which checks variants against the manifest but not group names or path safety. So the path-separator guard and the new duplicate-name check in `TestMatrix.Validate()` only actually run in `create-guests`, not in the binary that writes the JUnit files. A malformed `TEST_PLAN` (unsafe or duplicate `Name`) therefore panics inside the per-group worker goroutines (`g.JUnitFile()` at lines 96/119, no `recover`), tearing down the whole run mid-flight instead of failing fast. Calling `plan.Validate()` right after `ResolveTestPlan` would make both checks effective here and turn this into a clean up-front error. (The CI pipeline is mostly shielded today since create-guests validates first, but a standalone `run-tests` isn't.)

**Comment by mgencur** [comment]
@ironcladlou Thanks, looks great!
If I understand it correctly the user can specify a TestPlan but he has to choose a variant from the variants hardcoded in the platform's ClusterSpecs [here](https://github.com/openshift/hypershift/blob/main/test/e2e/v2/lifecycle/aws.go#L53).
What do you think about including clusterSpecs next to the testMatrix like this?
```
{
  "name": "upgrade",
  "platform": "azure",
  "clusterSpecs": {
    "variant": "xyz",
    "extraArgs": ...,
    "releaseImage": ...
  },
  "testMatrix": {
    "sequential": [
      {"name": "upgrade-flow", "steps": [
        {"name": "pre", "variant": "xyz", "labelFilter": "upgrade-pre"},
        {"name": "run", "variant": "xyz", "labelFilter": "upgrade-run"}
      ]}
    ]
  }
}
```
In this way, the user could define both the cluster spec and test matrix for specs that are not hardcoded in the binary. 
These can be specific configurations, which we don't want to have in the default cluster specs, or for ad-hoc testing.
The startup logic would be shared, just the parameters different.

**Comment by ironcladlou** [comment]
> @ironcladlou Thanks, looks great! If I understand it correctly the user can specify a TestPlan but he has to choose a variant from the variants hardcoded in the platform's ClusterSpecs [here](https://github.com/openshift/hypershift/blob/main/test/e2e/v2/lifecycle/aws.go#L53). What do you think about including clusterSpecs next to the testMatrix like this?
> 
> ```
> {
>   "name": "upgrade",
>   "platform": "azure",
>   "clusterSpecs": {
>     "variant": "xyz",
>     "extraArgs": ...,
>     "releaseImage": ...
>   },
>   "testMatrix": {
>     "sequential": [
>       {"name": "upgrade-flow", "steps": [
>         {"name": "pre", "variant": "xyz", "labelFilter": "upgrade-pre"},
>         {"name": "run", "variant": "xyz", "labelFilter": "upgrade-run"}
>       ]}
>     ]
>   }
> }
> ```
> 
> In this way, the user could define both the cluster spec and test matrix for specs that are not hardcoded in the binary. These can be specific configurations, which we don't want to have in the default cluster specs, or for ad-hoc testing. The startup logic would be shared, just the parameters different.

@mgencur I think we've been having similar thoughts along those lines, but there's some additional design work to think through for the cluster specs. The variants are coupled to imperative logic per platform (the hooks like post, etc.), and making them fully declarative will take some some design. I'm concerned that further leaning on extra-args will turn into a mess of argument strings spread everywhere, and I expect it won't be long before people want to customize the hook logic as well. I wonder if we could move forward with the current PR as an incremental improvement and then work on a followup which deals with cluster specs in a way that feels cohesive with the move towards a more declarative and composable design...

File: test/e2e/v2/lifecycle/aws.go (hunk 9 above)
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/aws.go:122]: I made the interface docs for DefaultTestPlan more precise (I think), as to introducing a stronger coupling between the matrix and the clusterspecs, I'd like to deal with that in a followup since I think we could start lifting out the specs into a more declarative form that will help address it more cleanly that I can really do here without increasing the scope of changes too far for now

File: test/e2e/v2/lifecycle/testplan.go (hunk 10 above)
Comment (ironcladlou) [reply test/e2e/v2/lifecycle/testplan.go]: I cleaned this up with single truth source

File: test/e2e/v2/cmd/run-tests/main.go (hunk 11 above)
Comment (ironcladlou) [reply test/e2e/v2/cmd/run-tests/main.go:74]: I cleaned this all up so that validation covers the path checks and makes the panic a backstop (i.e. if you panic it's because you forgot to call validate, which is explicitly required / a bug)

**Comment by ironcladlou** [comment]
Flaky unit test: https://redhat.atlassian.net/browse/OCPBUGS-114415

/test unit

**Comment by ironcladlou** [comment]
/verified by e2e regression analysis and manual testing of the new feature with the following test plan

```yaml
name: aws-public-nonlifecycle
platform: aws
testMatrix:
  parallel:
    - name: public
      variant: public
      labelFilter: "!lifecycle || hosted-cluster-aws"
```
